Hollow Knight updates (goals, WP/POP, etc.)

dewiniaid opened the pull request:
I don't know AP internals very well and haven't extensively tested this yet.  Posting as-is to request feedback.

This all assumes the client behavior if a location isn't randomized is to simply leave it alone.  I _think_ this is the case, but haven't checked yet.

This makes the following changes under `worlds/hk`:

- Add configurable goals (Any, THK, Siblings, Radiance)
  - Change base logic to require Opened_Black_Egg_Temple instead of requiring 3 dreamers.  This is future-proof for transition rando, where Black Egg might not have been located yet.
  - Add combat logic for THK and Radiance on par with Rando4's boss logic, so itemless HK shouldn't be required.
  - Completing a goal different than your assigned one is allowed, provided you meet the combat and item requirements.
    - This means you can win by killing Radiance even when your goal is THK (but the reverse is not true)
    - Godhome ending (i.e. Absolute Radiance) satisfies all victory conditions provided you also have the required items (e.g. 3 Dreamers and possibly Dream Nail and Void Heart)

- Add White Palace options (Exclude, King Fragment Only, No Path of Pain, Include)
  - Excluded WP may still be required for King Fragment if Charms are not randomized
  - Simply don't place WP locations that are excluded 
  - Distinguish between POP locations (required for POP), WP checks (actual item locations), WP transitions (relevant for future transition rando), and WP events (logically required to reach King Fragment)
  - Many transitions were listed twice.  Remove duplicates.
  - Sort transitions by scene

~~- For randomizable locations that have no logical significance when not randomized, simply skip adding them to the pool entirely for theoretically faster generation.~~

- Add support for random starting geo

- Add more sensible item group names for hinting the following: Dreamer, Fragments, Cloak, Claw, CDash, Horizontal, and Vertical

- Minor typo fixes in comments

dewiniaid commented:
I've confirmed the goals side of this actually works.  Generated spoiler logs _look_ accurate but need actual testing to be sure; I haven't played an actual game with this yet.  Nonetheless, changing this out of draft status.

Ijwu commented:
Ok there's been some work on this on the clientside in the meantime. @dewiniaid is this code still correct in terms of that?

dewiniaid commented:
Rebased against latest main.

Also added random start geo since it was free, and removed code that would skip generation of non-logically-significant vanilla locations (i.e. geo rocks when geo rocks are off) since it had negative implications for same-slot coop.

This PR in its current state matches the code used for the successful beta test a few hours ago.

Ijwu commented on worlds/hk/__init__.py:
```diff
@@ -149,23 +158,42 @@ def generate_early(self):
+            exclusions.update(path_of_pain_locations)
+        if wp <= WhitePalace.option_kingfragment:
+            exclusions.update(white_palace_checks)
+        if wp == WhitePalace.option_exclude and self.world.RandomizeCharms[self.player]:
+            # Ensure KF location is still reachable if charms are non-randomized
```
The comment on line 169 says to do something if charms are _non-randomized_ but the logic of the if does stuff if they _are randomized_. Is this intentional?

dewiniaid replied:
Took me a moment to untangle my own logic here, but this is correct:

If White Palace is excluded, we only actually go ahead and exclude the transitions/events/etc related to King Fragment if charms are randomized.  If they are not randomized, then King Fragment is required for most goals and needs to be logically reachable in e.g. ALR or beatable only accessibility -- so we do not exclude transitions or events in WP or the King Fragment location.

Ijwu commented on worlds/hk/__init__.py:
```diff
@@ -218,11 +247,58 @@ def create_items(self):
-            world.completion_condition[player] = lambda state: state.has('DREAMER', player, 3)
+            goal = world.Goal[player]
+            if goal == Goal.option_siblings:
+                world.completion_condition[player] = lambda state: (
+                    self._can_beat_thk(state, player)
+                    and state.has('WHITEFRAGMENT', player, 3)
+                )
```
This could be set to use the `HKLogicMixin` if you move the `_can_beat_thk()` to the mixin class.

dewiniaid replied:
This should now be done.  HKLogicMixin didn't exist when I first opened this, hopefully I used it correctly.

Ijwu commented on worlds/hk/__init__.py:
```diff
@@ -218,11 +247,58 @@ def create_items(self):
+                    and state.has('WHITEFRAGMENT', player, 3)
+                )
+            elif goal == Goal.option_radiance:
+                world.completion_condition[player] = lambda state: (
+                    state.count('Opened_Black_Egg_Temple', player)
+                    and state.has('WHITEFRAGMENT', player, 3)
+                    and state.has('DREAMNAIL', player, 1)
+                    and (
+                        (state.has('LEFTCLAW', player) and state.has('RIGHTCLAW', player))
+                        or state.has('WINGS', player)
+                    )
+                    and (
+                        state.count('FIREBALL', player) + state.count('SCREAM', player)
+                        + state.count('QUAKE', player)
+                    ) > 1
+                    and (  # NAILCOMBAT
+                        state.count('LEFTSLASH', player)
+                        or state.count('RIGHTSLASH', player)
+                        or state.count('UPSLASH', player)
+                    )
+                    and (
+                        (state.has('LEFTDASH', player, 2) and state.has('RIGHTDASH', player, 2))  # Both Shade Cloaks
+                        or (self.world.ProficientCombat[self.player] and state.has('QUAKE', player))  # or Dive
+                    )
+                )
```
Whatever logic calculations need to happen here should be in `HKLogicMixin` if possible then referred to here.

dewiniaid replied:
This should now be done.  HKLogicMixin didn't exist when I first opened this, hopefully I used it correctly.

Ijwu commented on worlds/hk/__init__.py:
```diff
@@ -218,11 +247,58 @@ def create_items(self):
             for loc, price in zip(locations, prices):
                 loc.cost = price
 
+    def _can_beat_thk(self, state, player):
```
Does this belong here? Shouldn't this be a part of `HKLogicMixin`?

dewiniaid replied:
This should now be done.  HKLogicMixin didn't exist when I first opened this, hopefully I used it correctly.

Ijwu commented:
Added @Berserker66 as reviewer because he does know this side of things better than me.

Berserker66 commented:
Looked over it, I see the same thing Ijwu does, would be more in-line with other implementations if LogicMixin was used.

Ijwu reviewed (approved):
Seems good, Will let @Berserker66 merge so he has a chance to double check too,

dewiniaid commented:
Updated to use HKLogicMixin.  Hopefully I did that correctly, though the playthrough in a few test seeds I rolled looks correct.

dewiniaid commented:
Updated item_name_groups for more friendly !hint usage.  Edited first message with a complete list of what is now actually changed.

Longer term planning for HK includes potentially renaming items to have more friendly names ("Progressive Cloak", etc.) which will simplify some of this, but this is a good interim option.
